### ISSUE TITLE
MetaCognition Agent steering

### ISSUE THREAD

**Issue by Tyler-R-Kendrick** [Tyler-R-Kendrick/agentic-metacognition#4]
Use TDD to build an agent that leverages the persistent steering data.

---

Build it as a **hybrid agent**:

* a **planner/judge** that decides when extra reasoning or retrieval is needed
* a **steered executor** running on an **open-weights model** where you can actually read hidden states, encode them with an SAE, persist sparse features, and inject steering vectors during generation

That split matters because OpenAI’s hosted reasoning models expose useful reasoning controls and summaries, but not raw hidden states or raw reasoning tokens through the API. The Responses API gives you reasoning controls such as `reasoning.effort`, stateful turns, encrypted reasoning items, and summaries, but the raw internal reasoning tokens are not exposed. ([[OpenAI Developers](https://developers.openai.com/api/docs/guides/reasoning/)][1])

## 1. Choose the right architecture

Use this topology:

**User / task**
→ **Planner** (hosted reasoning model)
→ **Retriever + tool router**
→ **Steered local model** (open weights, hooks enabled)
→ **Verifier / judge**
→ **Memory write-back**

Why this topology:

* OpenAI reasoning models are well-suited for complex planning and multi-step agent workflows. OpenAI explicitly recommends mixing reasoning models for planning with faster models for execution depending on cost and accuracy needs. ([[OpenAI Developers](https://developers.openai.com/api/docs/guides/reasoning/)][1])
* SAE steering requires access to hidden activations or layer outputs. Hugging Face Transformers supports `output_hidden_states=True`, which is the practical path for building the steered execution layer. ([[Hugging Face](https://huggingface.co/docs/transformers/en/main_classes/output?utm_source=chatgpt.com)][2])

## 2. Core design principle

Do **not** try to directly “steer reasoning” from a single contrastive vector and assume it is causal.

Recent papers show both the promise and the fragility here:

* SAE-based steering can produce controllable reasoning-like behavior and can outperform simple activation-difference steering on some reasoning benchmarks. ([[arXiv](https://arxiv.org/pdf/2510.01246)][3])
* But a 2026 falsification study found that many purported “reasoning features” were actually lexical or superficial correlates, and steering them gave minimal or slightly negative changes in benchmark performance. ([[arXiv](https://arxiv.org/abs/2601.05679?utm_source=chatgpt.com)][4])

So the right production stance is:

**treat sparse features as candidate control handles, not ground-truth reasoning circuits.**

## 3. What to persist

Persist three classes of objects.

### A. Sparse feature traces

For each run, save:

* model id + revision
* prompt/task type
* layer index
* token positions
* top-k SAE latent activations per token
* final outcome labels: success, hallucination, overthinking, tool misuse, refusal mismatch, etc.

This gives you the raw material for deriving steering vectors later.

### B. Feature prototypes

For each desired behavioral mode, persist:

* positive centroid in sparse latent space
* negative centroid
* contrast vector = `mu_pos - mu_neg`
* sparsified mask of stable dimensions
* confidence stats by task family

Examples of modes:

* careful decomposition
* citation-seeking
* tool-before-answer
* code-review mode
* concise direct-answer mode
* anti-loop / anti-meander

### C. Intervention policies

Persist rules like:

* apply feature set F at layers 18, 22, 26
* decay coefficient by token step
* disable after tool observation
* only apply if verifier confidence < threshold
* block if repetition score rises

This is critical because later SAE-steering work reports that **constant steering** can create degenerate repetitive outputs, while **token-wise decaying steering** works better. ([[arXiv](https://arxiv.org/pdf/2510.01246)][3])

### Step 1: collect activations

Run an open-weights base model on a task corpus that matches your agent’s workload:

* QA with retrieval
* tool-using tasks
* long-form synthesis
* coding tasks
* your domain tasks

Capture hidden states from a small set of candidate layers first, not every layer. Mid-to-late residual stream layers are the usual starting point for steering experiments, though the exact best layer is empirical. The SAE literature is built around decomposing hidden states into sparse overcomplete representations. ([[arXiv](https://arxiv.org/html/2505.16188v2)][5])

### Step 2: train SAEs per target layer

For each chosen layer:

* input: residual stream activation
* encoder: sparse projection
* decoder: reconstruction
* loss: reconstruction + sparsity penalty

Store:

* encoder weights
* decoder weights
* feature norms
* dead-feature statistics
* layer-wise reconstruction error

### Step 3: attach interpretation metadata

For each latent feature, attach:

* highest-activating task examples
* top triggering token windows
* correlation with outcome labels
* whether it is stable across paraphrases
* whether it survives counterfactual token edits

This last step matters because the falsification paper shows many “reasoning” features vanish under token-level stress tests or are explainable by shallow lexical triggers. ([[arXiv](https://arxiv.org/abs/2601.05679?utm_source=chatgpt.com)][4])

## 5. Deriving steering vectors

Use a **three-stage pipeline**.

### Stage A: candidate retrieval

Start with contrastive sets:

* successful careful reasoning vs shallow failures
* grounded tool use vs unsupported answer-first behavior
* concise correct vs verbose meandering
* policy-compliant vs policy-violating trajectories

Map hidden states through the SAE encoder and compute:

* latent activation differences
* mutual information with outcome labels
* stability under paraphrase

This is aligned with SAE-Steering style work that first narrows a huge feature pool to a small candidate set. ([[arXiv](https://arxiv.org/pdf/2601.03595)][6])

### Stage B: feature ranking

Rank candidate sparse features by:

* causal impact under intervention
* low lexical confounding
* cross-task consistency
* low collateral damage on unrelated tasks

Use ablations:

* inject only feature i
* suppress only feature i
* inject top-1
* inject top-k
* compare constant vs decayed schedules

Recent comparative work suggests that a **single strong latent** or a very small sparse set can outperform broader, redundant steering, but only if the schedule is controlled well. ([[arXiv](https://arxiv.org/pdf/2510.01246)][3])

### Stage C: bundle into named controllers

Turn raw features into named controls:

* `deliberate_plan_v1`
* `retrieve_before_commit_v2`
* `anti_repetition_v1`
* `strict_citation_v1`
* `tool_error_recovery_v1`

Each controller should contain:

* latent ids
* coefficients
* layers
* schedule
* activation trigger conditions
* validation metrics

## 6. Intervention mechanism at inference time

At each decoding step on the local model:

1. capture the current layer activation
2. encode into SAE latent space
3. optionally log current sparse code
4. add or suppress selected sparse dimensions
5. decode back into residual-space delta
6. inject delta into the residual stream
7. continue generation

Use:

`h' = h + alpha_t * D(m ⊙ (z_target - z_current_baseline))`

where:

* `h` = current hidden state
* `z` = SAE latent code
* `m` = sparse mask over chosen latent dimensions
* `D` = SAE decoder
* `alpha_t` = token-wise schedule

In practice, keep `alpha_t` decaying with token index unless evaluation proves otherwise. Constant injection is more likely to cause degeneration. ([[arXiv](https://arxiv.org/pdf/2510.01246)][3])

## 7. Agent loop that leverages the steering

A good production loop looks like this:

### Planner

Use a reasoning model to decide:

* task type
* whether retrieval/tools are needed
* whether to invoke the steered local model
* which steering controller to request

Hosted reasoning models support effort tuning and can be used as planners for complex multi-step workflows. ([[OpenAI Developers](https://developers.openai.com/api/docs/guides/reasoning/)][1])

### Controller selector

Map planner output + task telemetry to a control profile.

Example:

* ambiguous scientific question → `deliberate_plan_v1 + strict_citation_v1`
* coding fix with stack trace → `tool_error_recovery_v1`
* customer support response → `concise_direct_v1`
* long answer with evidence gaps → `retrieve_before_commit_v2`

### Steered executor

Run the local open-weights model with:

* retrieval context
* task prompt
* controller payload
* layer hooks enabled

### Verifier

Check:

* factual grounding
* repetition/degeneration
* tool alignment
* policy adherence
* answer completeness

### Memory update

Persist:

* selected controller
* sparse traces
* verifier score
* final outcome
* any regression tags

That lets the system improve its controller catalog over time.

## 8. Retrieval and vector memory design

When you said “persisted vector extractions on activation space,” I would separate memory into two stores.

### Semantic store

A standard text embedding store for documents, tool traces, and prior tasks.

Use this for normal RAG.

### Activation-feature store

A dedicated store keyed by:

* task family
* model + layer
* controller id
* sparse feature ids
* outcome labels

Do **not** treat this like a generic embedding DB alone. You want structured lookup such as:

* “best controller for legal-argument synthesis on model X”
* “top sparse features correlated with unsupported claims”
* “feature bundles that improved retrieval-grounded coding tasks”

A plain vector DB is fine for nearest-neighbor lookup on feature centroids, but add relational metadata around it.

## 9. Evaluation protocol

Use four eval groups.

### Behavioral gain

Did steering improve:

* correctness
* groundedness
* tool-use quality
* policy compliance
* latency-adjusted utility

### Causal specificity

Does steering only affect the intended behavior?

You want the smallest shift that improves the target metric without broadly distorting output style.

### Robustness

Test against:

* paraphrases
* adversarial lexical triggers
* different document orderings
* tool failures
* long contexts

This is mandatory because sparse “reasoning features” may really be token-pattern features. ([[arXiv](https://arxiv.org/abs/2601.05679?utm_source=chatgpt.com)][4])

### Degeneration

Track:

* repetition
* mode collapse
* excessive verbosity
* answer refusal drift
* invalid tool call loops

This is especially important for constant steering schedules. ([[arXiv](https://arxiv.org/pdf/2510.01246)][3])

### Planner / judge

Use OpenAI Responses API with a reasoning model for planning and adjudication. The API supports reasoning effort, summaries, and stateful multi-turn workflows. ([[OpenAI Developers](https://developers.openai.com/api/docs/guides/reasoning/)][1])

### Steered executor

Use an open-weights Transformers model with hookable hidden states. Hugging Face documents hidden-state outputs and generation-time access patterns. ([[Hugging Face](https://huggingface.co/docs/transformers/en/main_classes/output?utm_source=chatgpt.com)][2])

### SAE training / intervention

PyTorch + custom hook layer.

### Storage

Use:

* Postgres for metadata and evals
* object storage for traces
* vector DB for controller centroids / nearest-neighbor retrieval

## 11. Minimal build order

Build in this order.

### Phase 1

Plain agent first:

* planner
* retriever
* local executor
* verifier
* telemetry

No steering yet.

### Phase 2

Add read-only activation logging:

* hidden-state capture
* SAE training
* sparse trace persistence
* analysis notebooks

### Phase 3

Offline intervention experiments:

* one layer
* one task family
* top-1 or top-k sparse features
* small coefficient sweep
* decayed schedule only

### Phase 4

Online gated steering:

* only on low-confidence tasks
* only with fallback to unsteered generation
* automatic verifier rollback

### Phase 5

Adaptive controller selection:

* learned routing based on task + prior success rates

## 12. A concrete control policy to start with

A good first controller is not “reason harder.”

Start with:

**`retrieve_before_commit_v1`**

Target behavior:

* do not answer substantive factual claims until evidence is grounded
* prefer tool or retrieval usage when support is missing
* reduce unsupported confident text

Why start there:

* easier to evaluate than abstract “reasoning quality”
* more operationally useful
* lower risk of lexical confounding than “show step-by-step reasoning”

## 13. Failure modes to expect

Expect these early:

* lexical-feature steering instead of real reasoning control ([[arXiv](https://arxiv.org/abs/2601.05679?utm_source=chatgpt.com)][4])
* repetition or degenerate loops from over-strong coefficients ([[arXiv](https://arxiv.org/pdf/2510.01246)][3])
* layer mismatch between training-time and inference-time hooks
* controller overfitting to one prompt style
* apparent gains that disappear under paraphrase or tool perturbation

## 14. Practical pseudocode

```python
# planner chooses controller
plan = planner.run(task, reasoning_effort="medium")

controller = controller_router.select(
    task_type=plan.task_type,
    needs_retrieval=plan.needs_retrieval,
    risk_level=plan.risk_level
)

context = retriever.get_context(task) if plan.needs_retrieval else []

# steered local execution
with steering_hooks(
    model=local_model,
    sae=sae_layer_22,
    controller=controller,   # latent ids, coeffs, schedule, layer
    schedule="decay"
):
    draft = local_model.generate(
        prompt=build_prompt(task, context, plan),
        output_hidden_states=True
    )

verdict = verifier.run(task, draft, context)

if verdict.fail:
    # fallback: weaker or no steering
    draft = fallback_generate(task, context)

memory.write({
    "task": task,
    "controller": controller.id,
    "verdict": verdict,
    "feature_trace_uri": trace_uri,
})
```

## 15. The most important recommendation

Use **OpenAI-hosted reasoning for planning/judgment** and a **local open-weights model for activation steering**.

That is the cleanest way to get both:

* robust agent planning and tool orchestration via the Responses API ([[OpenAI Developers](https://developers.openai.com/api/docs/guides/reasoning/)][1])
* direct hidden-state access for SAE encoding and intervention via an open model stack ([[Hugging Face](https://huggingface.co/docs/transformers/en/main_classes/output?utm_source=chatgpt.com)][2])

Trying to do SAE-style activation steering directly on hosted closed APIs is the wrong place to start because the necessary hidden-state control surface is not exposed. ([[OpenAI Developers](https://developers.openai.com/api/docs/guides/reasoning/)][1])

References worth keeping open:

* OpenAI reasoning models guide. ([[OpenAI Developers](https://developers.openai.com/api/docs/guides/reasoning/)][1])
* OpenAI reasoning best practices. ([[OpenAI Developers](https://developers.openai.com/api/docs/guides/reasoning-best-practices/)][7])
* SAE-Steering / controllable reasoning via sparse autoencoders. ([[arXiv](https://arxiv.org/pdf/2601.03595)][6])
* Comparative analysis of SAE vs activation-difference steering. ([[arXiv](https://arxiv.org/pdf/2510.01246)][3])
* SAE-SSV supervised steering in sparse representation spaces. ([[arXiv](https://arxiv.org/html/2505.16188v2)][5])
* Falsifying sparse autoencoder reasoning features. ([[arXiv](https://arxiv.org/abs/2601.05679?utm_source=chatgpt.com)][4])
* Hugging Face hidden-state outputs. ([[Hugging Face](https://huggingface.co/docs/transformers/en/main_classes/output?utm_source=chatgpt.com)][2])

I can turn this into a concrete repo skeleton next: service boundaries, data schemas, hook interfaces, and an eval harness.

[1]: https://developers.openai.com/api/docs/guides/reasoning/ "Reasoning models | OpenAI API"
[2]: https://huggingface.co/docs/transformers/en/main_classes/output?utm_source=chatgpt.com "Model outputs"
[3]: https://arxiv.org/pdf/2510.01246 "A Comparative Analysis of Sparse Autoencoder and Activation Difference in Language Model Steering"
[4]: https://arxiv.org/abs/2601.05679?utm_source=chatgpt.com "Do Sparse Autoencoders Identify Reasoning Features in Language Models?"
[5]: https://arxiv.org/html/2505.16188v2 "SAE-SSV: Supervised Steering in Sparse Representation Spaces for Reliable Control of Language Models"
[6]: https://arxiv.org/pdf/2601.03595 "Controllable LLM Reasoning via Sparse Autoencoder-Based Steering"
[7]: https://developers.openai.com/api/docs/guides/reasoning-best-practices/ "Reasoning best practices | OpenAI API"

---

Implement all phases using TDD. Implement all phases to completion. Don't stop until all phases are complete.
